crates/solver/src/driver/solver_settlements.rs

```diff
@@ -90,30 +93,144 @@ fn merge_at_most_settlements(
-) {
-    let settle_orders_older_than =
-        chrono::offset::Utc::now() - chrono::Duration::from_std(min_order_age).unwrap();
-    settlements.retain(|settlement| {
-        settlement
-            .trades()
-            .iter()
-            .any(|trade| trade.order.order_meta_data.creation_date <= settle_orders_older_than)
-    });
+/// Filters out all settlements without any orders which are mature by age or mature by association.
+/// Any order older than `min_order_age` is considered to be mature by age.
+/// Any younger order in a settlement containing an order mature by age or mature by association
+/// is considered to be mature by association.
+/// Because maturity by association is defined recursively it can "spread" across settlements,
+/// resulting in settlements being allowed where it's not immediately obvious by which association
+/// any order of a settlement has matured.
+pub fn filter_immature_settlements(min_order_age: Duration, settlements: &mut Vec<Settlement>) {
+    fn find_mature_settlements(
+        min_order_age: Duration,
+        settlements: &[Settlement],
+    ) -> HashSet<*const Settlement> {
+        let settle_orders_older_than =
+            chrono::offset::Utc::now() - chrono::Duration::from_std(min_order_age).unwrap();
+
+        let mut valid_trades = HashSet::<&model::order::Order>::default();
+        // Use *const T to avoid expensive clones and hashing. This also doesn't require T: Hash + Eq.
+        let mut valid_settlements = HashSet::<*const Settlement>::default();
+
+        loop {
+            let mut new_order_added = false;
+
+            for settlement in settlements.iter() {
+                if valid_settlements.contains(&(settlement as _)) {
+                    break;
+                }
+
+                let contains_valid_trade = settlement.trades().iter().any(|trade| {
+                    // mature by age
+                    trade.order.order_meta_data.creation_date <= settle_orders_older_than
+                    // mature by association
+                    || valid_trades.contains(&trade.order)
+                });
+
+                if contains_valid_trade {
+                    for trade in settlement.trades() {
+                        // make all orders within this settlement mature by association
+                        new_order_added |= valid_trades.insert(&trade.order);
+                    }
+                    valid_settlements.insert(settlement);
+                }
+            }
+
+            if !new_order_added {
+                break valid_settlements;
+            }
+        }
+    }
+
+    let valid_settlements = find_mature_settlements(min_order_age, &settlements[..]);
+    // After finding mature settlements DON'T mutate `settlements` other than filtering them with
+    // `valid_settlements`. Everything else might invalidate contents of `valid_settlements`!
+    settlements.retain(|settlement| valid_settlements.contains(&(settlement as _)))
```

Comment:
Hmm, I just thought about this now, but this is dependent on the specific implementation of `retain` which is not allow to move the inner `settlements` to new locations in memory temporarily. `Vec::retain` doesn't guarantee that the elements are pinned.

Although its less efficient (requires allocation, etc.) maybe it makes more sense to do this filtering on indices and construct a new `Vec<Settlement>` for the filtered result?

Reply:
Yeah I checked that this isn't a problem with the current implementation of `retain` but I didn't think about that it might change in the future.
I now went with your idea of filtering based on indices and collecting into a new `Vec<Settlement>`.